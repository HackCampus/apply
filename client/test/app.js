const test = require('ava')
const sinon = require('sinon')
const pull = require('pull-stream')

const app = require('../app')

// emits values given in the array after the specified timeouts.
// data : [[timeout, value]]
function timedSource(data) {
  return pull(
    pull.values(data),
    pull.asyncMap(function(item, cb) {
      setTimeout(function() {
        cb(null, item[1])
      }, item[0]);
    })
  )
}

const action = (type, payload) => ({type, payload})

test('init', t => {
  t.snapshot(app.init())
})

test('view', t => {
  const {model} = app.init()
  const dispatch = sinon.stub()
  const view = app.view(model, dispatch)
  t.snapshot(view.toString())
  t.false(dispatch.called)
})

test.cb('autosave - ignore everything other than application updates', t => {
  const actions = () => pull.values([action('ignore me')])
  const sources = {actions}
  pull(
    app.run(action('autosave'), sources),
    pull.collect((err, saveActions) => {
      t.deepEqual(saveActions, [])
      t.end()
    })
  )
})

test.cb('autosave - emit save when application changes', t => {
  const actions = () => pull.values([{child: 'personalDetails', action: action('whatever')}])
  const sources = {actions}
  pull(
    app.run(action('autosave'), sources),
    pull.collect((err, saveActions) => {
      t.deepEqual(saveActions, [action('saveApplication')])
      t.end()
    })
  )
})

test.cb('autosave - debounce', t => {
  const actions = () => timedSource([
    [0, {child: 'personalDetails', action: action('whatever')}],
    [100, {child: 'questions', action: action('whatever')}],
    [200, {child: 'techPreferences', action: action('whatever')}],
  ])
  const sources = {actions}
  pull(
    app.run(action('autosave'), sources),
    pull.collect((err, saveActions) => {
      t.deepEqual(saveActions, [action('saveApplication')])
      t.end()
    })
  )
})

test.cb('autosave - save again after 3 seconds', t => {
  const actions = () => timedSource([
    [0, {child: 'personalDetails', action: action('whatever')}],
    [100, {child: 'questions', action: action('whatever')}],
    [200, {child: 'techPreferences', action: action('whatever')}],
    [3500, {child: 'techPreferences', action: action('whatever')}],
  ])
  const sources = {actions}
  pull(
    app.run(action('autosave'), sources),
    pull.collect((err, saveActions) => {
      t.deepEqual(saveActions, [action('saveApplication'), action('saveApplication')])
      t.end()
    })
  )
})
