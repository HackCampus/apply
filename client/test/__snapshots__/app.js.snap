exports[`init 1`] = `
Object {
  "effect": Array [
    Object {
      "payload": undefined,
      "type": "waitForUser",
    },
    Object {
      "child": "authenticate",
      "effect": Array [
        Object {
          "payload": undefined,
          "type": "loadUser",
        },
      ],
    },
  ],
  "model": Object {
    "application": null,
    "children": Object {
      "authenticate": Object {
        "children": Object {
          "confirmPassword": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "email": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "password": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
        },
        "error": null,
        "password": false,
        "tab": 0,
        "user": null,
      },
      "finishApplication": Object {
        "children": Object {},
      },
      "personalDetails": Object {
        "children": Object {
          "contactEmail": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "courseName": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "courseType": Object {
            "choices": Array [
              "under-graduate",
              "graduate",
              "other",
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "cvUrl": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "dateOfBirth": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "firstName": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "gender": Object {
            "choices": Array [
              "male",
              "female",
              "other",
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "graduationYear": Object {
            "choices": Array [
              "2017",
              "2018",
              "2019",
              "2020",
              "2021",
              "other",
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "lastName": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "otherCourseType": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "otherGraduationYear": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "otherUniversity": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "otherYearOfStudy": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "referer": Object {
            "choices": Array [
              "friend",
              "hackathon",
              "campus event",
              "Facebook",
              "other",
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "refererDetail": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "university": Object {
            "options": Array [
              "other (eg. international)",
              "Anglia Ruskin University",
              "Arts University Bournemouth",
              "University of the Arts London",
              "Camberwell College of Arts",
              "Central Saint Martins College of Art and Design",
              "Chelsea College of Art and Design",
              "London College of Communication",
              "London College of Fashion",
              "Wimbledon College of Art",
              "Aston University",
              "University of Bath",
              "Bath Spa University",
              "University of Bedfordshire",
              "University of Birmingham",
              "Birmingham City University",
              "Laedge International University",
              "University College Birmingham",
              "Bishop Grosseteste University",
              "University of Bolton",
              "Bournemouth University",
              "BPP University",
              "University of Bradford",
              "University of Brighton",
              "University of Bristol",
              "Brunel University",
              "University of Buckingham",
              "Buckinghamshire New University",
              "University of Cambridge",
              "Canterbury Christ Church University",
              "University of Central Lancashire",
              "University of Chester",
              "University of Chichester",
              "City University London",
              "Coventry University",
              "Cranfield University",
              "University for the Creative Arts",
              "University of Cumbria",
              "De Montfort University",
              "University of Derby",
              "Durham University",
              "University of East Anglia",
              "University of East London",
              "Edge Hill University",
              "University of Essex",
              "University of Exeter",
              "Falmouth University",
              "University of Gloucestershire",
              "University of Greenwich",
              "Harper Adams University",
              "University of Hertfordshire",
              "University of Huddersfield",
              "University of Hull",
              "Imperial College London",
              "Keele University",
              "University of Kent",
              "Kingston University",
              "Lancaster University",
              "University of Leeds",
              "Leeds Metropolitan University",
              "Leeds Trinity University",
              "University of Leicester",
              "University of Lincoln",
              "University of Liverpool",
              "Liverpool Hope University",
              "Liverpool John Moores University",
              "University of London",
              "Birkbeck, University of London",
              "Courtauld Institute of Art",
              "Goldsmiths, University of London",
              "Heythrop College",
              "Institute of Cancer Research",
              "Institute of Education",
              "King\'s College London",
              "London Business School",
              "London School of Economics",
              "London School of Hygiene & Tropical Medicine",
              "Queen Mary, University of London",
              "Royal Northern College of Music",
              "Royal Academy of Music",
              "Royal Holloway, University of London",
              "Royal Veterinary College",
              "St George\'s, University of London",
              "School of Advanced Study",
              "Institute of Advanced Legal Studies",
              "Institute of Classical Studies",
              "Institute of Commonwealth Studies",
              "Institute of English Studies",
              "Institute of Germanic and Romance Studies",
              "Institute of Historical Research",
              "Institute for the Study of the Americas",
              "Warburg Institute",
              "School of Oriental and African Studies",
              "School of Pharmacy, University of London",
              "University College London",
              "University Marine Biological Station Millport",
              "University of London Institute in Paris",
              "London Metropolitan University",
              "London South Bank University",
              "Loughborough University",
              "University of Manchester",
              "Manchester Metropolitan University",
              "Middlesex University",
              "Newcastle University",
              "Newman University",
              "University of Northampton",
              "Northumbria University",
              "Norwich University of the Arts",
              "University of Nottingham",
              "Nottingham Trent University",
              "Open University",
              "University of Oxford",
              "Oxford Brookes University",
              "University of Plymouth",
              "University of Portsmouth",
              "University of Reading",
              "Richmond University",
              "Roehampton University",
              "Royal Agricultural University",
              "University of Salford",
              "University of Sheffield",
              "Sheffield Hallam University",
              "University of Southampton",
              "Southampton Solent University",
              "University of St Mark & St John",
              "St. Mary\'s University College",
              "Staffordshire University",
              "University of Sunderland",
              "University of Surrey",
              "University of Sussex",
              "Teesside University",
              "Thames Valley University",
              "University of Warwick",
              "University of Westminster",
              "University of the West of England",
              "University of Winchester",
              "University of Wolverhampton",
              "University of Worcester",
              "University of York",
              "York St John University",
              "University of Aberdeen",
              "University of Abertay",
              "Aberystwyth University",
              "American InterContinental University London",
              "Bangor University",
              "British College of Osteopathic Medicine",
              "British School of Osteopathy",
              "Bucks New University",
              "Cardiff University",
              "Cardiff University of Wales Institute",
              "Cass Business School",
              "Central School of Speech and Drama",
              "Conservatoire for Dance and Drama",
              "Dartington College of Arts",
              "University of Dundee",
              "University of Edinburgh",
              "Edinburgh College of Art",
              "European School of Osteopathy",
              "University of Glamorgan",
              "Falmouth University College",
              "Glasgow Caledonian University",
              "Glasgow School of Art",
              "University of Glasgow",
              "Guildhall School of Music and Drama",
              "Henley Management College",
              "Heriot-Watt University",
              "Institute for System Level Integration",
              "Institute of Germanic & Romance Studies",
              "Lampeter University of Wales",
              "Leeds College of Music",
              "Leeds Trinity and All Saints",
              "London School of Economics and Political Science",
              "London School of Hygiene and Tropical Medicine",
              "Manchester Business School",
              "Napier University",
              "New College Nottingham",
              "Newman University College Birmingham",
              "Newport University of Wales",
              "North East Wales Institute of Higher Education",
              "Northern School of Contemporary Dance",
              "Queen Margaret University",
              "Queen\'s University Belfast",
              "Ravensbourne College of Design and Communication",
              "RCN Institute",
              "Regent\'s College",
              "Robert Gordon University",
              "Rose Bruford College",
              "Royal Agricultural College",
              "Royal College of Art",
              "Royal Scottish Academy of Music and Drama",
              "Royal Welsh College of Music and Drama",
              "Scarborough Campus University of Hull",
              "Scottish Agricultural College",
              "University of St Andrews",
              "St Mary\'s College",
              "St Mary\'s University College",
              "University of Stirling",
              "Stranmillis University College",
              "University of Strathclyde",
              "Swansea Institute of Higher Education",
              "Swansea University",
              "Trinity College Carmarthen",
              "UHI Millennium Institute",
              "Trinity Laban",
              "University of Ulster",
              "University College Plymouth St Mark & St John",
              "University of Exeter in Cornwall",
              "University of Wales",
              "University of the West of Scotland",
              "Writtle College",
            ],
            "selected": -1,
            "started": false,
            "value": null,
          },
          "websiteUrl": Object {
            "errors": Array [],
            "started": false,
            "value": "",
          },
          "yearOfStudy": Object {
            "choices": Array [
              "1",
              "2",
              "3",
              "4",
              "5",
              "other",
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
        },
        "errorFields": Object {},
        "errorMessage": null,
        "readOnly": false,
      },
      "questions": Object {
        "children": Object {
          "bestProject": Object {
            "started": false,
            "value": "",
          },
          "codeReview": Object {
            "started": false,
            "value": "",
          },
          "implementation": Object {
            "started": false,
            "value": "",
          },
          "mostExcitingTechnology": Object {
            "started": false,
            "value": "",
          },
          "perfectRole": Object {
            "started": false,
            "value": "",
          },
        },
      },
      "techPreferences": Object {
        "children": Object {
          "Android": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Angular": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "C#": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "CSS": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Django": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Docker": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "ElasticSearch": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Go": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Groovy / Grails": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Haskell": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Java": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "JavaScript": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Laravel": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "MongoDB": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "MySQL": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Neo4J": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Node.js": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Objective-C": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "PHP": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "PostgreSQL": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Python": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "RabbitMQ": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Rails": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "React": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Redis": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Ruby": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Scala": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "Swift": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
          "iOS": Object {
            "choices": Array [
              0,
              1,
              2,
              3,
            ],
            "chosen": -1,
            "started": false,
            "value": null,
          },
        },
      },
    },
    "errorFields": Object {},
    "errorMessage": null,
    "readOnly": false,
    "user": null,
  },
}
`;
